File: prompt_compressor/src/lexer.rs
use bitflags::bitflags;
use std::str::CharIndices;

bitflags! {
    /// Flags to annotate tokens during lexing or later optimization passes.
    #[derive(Debug, Clone, Copy, PartialEq, Eq)]
    pub struct TokenFlags: u32 {
        const NONE = 0;
        const CAPITALIZED = 1 << 0;
        const ALL_CAPS = 1 << 1;
        const HAS_NUMBERS = 1 << 2;
        const IS_STOP_WORD = 1 << 3;
        const IS_BOILERPLATE = 1 << 4;
        /// Set on tokens that a compression pass *synthesized* (e.g. a
        /// single collapsed space standing in for a run of whitespace)
        /// rather than sliced directly from the source text. Serialization
        /// uses this to know whether `start..end` is a real byte range to
        /// slice, or just a placeholder.
        const SYNTHETIC = 1 << 5;
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenKind {
    Word,
    Number,
    String,
    Url,
    CodeBlock,
    InlineCode,
    Punctuation,
    Operator,
    Bracket,
    Whitespace,
    Newline,
    Comment,
    Unknown,
}

/// A lexed token. Deliberately holds no owned text -- only the byte range
/// into the original source string. This makes `Token` a plain `Copy`
/// struct (no heap allocation, no allocator overhead per token), which is
/// the single biggest lever for staying inside a tight memory budget: the
/// original design allocated a fresh `String` per token and then cloned
/// that string multiple times per compression pass.
#[derive(Debug, Clone, Copy)]
pub struct Token {
    pub kind: TokenKind,
    pub start: usize,
    pub end: usize,
    pub flags: TokenFlags,
}

impl Token {
    /// Borrow this token's text out of `source`. O(1), zero allocation.
    /// `source` must be the same string the token was produced from (or,
    /// for SYNTHETIC tokens, the caller shouldn't call this at all --
    /// see `TokenStream::serialize`).
    #[inline]
    pub fn text<'a>(&self, source: &'a str) -> &'a str {
        &source[self.start..self.end]
    }
}

/// The Lexer acts as the Front-End scanner for our prompt compiler.
pub struct Lexer<'a> {
    input: &'a str,
    chars: CharIndices<'a>,
    current_char: Option<(usize, char)>,
    prev_char: Option<char>,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        let mut chars = input.char_indices();
        let current_char = chars.next();
        Self {
            input,
            chars,
            current_char,
            prev_char: None,
        }
    }

    /// Consumes the entire input and returns a vector of tokens.
    pub fn lex_all(mut self) -> Vec<Token> {
        let mut tokens = Vec::new();
        while let Some(token) = self.next_token() {
            tokens.push(token);
        }
        tokens
    }

    fn advance(&mut self) {
        self.prev_char = self.current_char.map(|(_, c)| c);
        self.current_char = self.chars.next();
    }

    fn peek_char(&self) -> Option<char> {
        self.current_char.map(|(_, c)| c)
    }

    fn current_pos(&self) -> usize {
        self.current_char.map(|(i, _)| i).unwrap_or(self.input.len())
    }

    fn starts_with(&self, prefix: &str) -> bool {
        self.input[self.current_pos()..].starts_with(prefix)
    }

    pub fn next_token(&mut self) -> Option<Token> {
        let (start, c) = self.current_char?;

        if c == '\n' || c == '\r' {
            return Some(self.lex_newline(start));
        }
        if c.is_whitespace() {
            return Some(self.lex_whitespace(start));
        }

        if c == '`' {
            if self.starts_with("```") {
                return Some(self.lex_code_block(start));
            } else {
                return Some(self.lex_inline_code(start));
            }
        }

        if "()[]{}<>".contains(c) {
            self.advance();
            return Some(self.create_token(TokenKind::Bracket, start, self.current_pos(), TokenFlags::NONE));
        }

        if ".,;:!?".contains(c) {
            self.advance();
            return Some(self.create_token(TokenKind::Punctuation, start, self.current_pos(), TokenFlags::NONE));
        }

        if c == '/' {
            if self.starts_with("//") {
                return Some(self.lex_line_comment(start));
            } else if self.starts_with("/*") {
                return Some(self.lex_block_comment(start));
            }
        }

        if "+-*/=~|&^%\\".contains(c) {
            self.advance();
            return Some(self.create_token(TokenKind::Operator, start, self.current_pos(), TokenFlags::NONE));
        }

        if c == '"' {
            return Some(self.lex_string(start, c));
        }

        if c == '\'' {
            let prev_is_alnum = self.prev_char.map_or(false, |pc| pc.is_alphanumeric());
            if prev_is_alnum {
                self.advance();
                return Some(self.create_token(TokenKind::Punctuation, start, self.current_pos(), TokenFlags::NONE));
            }
            return Some(self.lex_string(start, c));
        }

        if c.is_ascii_digit() {
            return Some(self.lex_number(start));
        }

        if c.is_alphabetic() || c == '_' {
            if self.starts_with("http://") || self.starts_with("https://") {
                return Some(self.lex_url(start));
            }
            return Some(self.lex_word(start));
        }

        self.advance();
        Some(self.create_token(TokenKind::Unknown, start, self.current_pos(), TokenFlags::NONE))
    }

    fn lex_newline(&mut self, start: usize) -> Token {
        while let Some(c) = self.peek_char() {
            if c == '\n' || c == '\r' {
                self.advance();
            } else {
                break;
            }
        }
        self.create_token(TokenKind::Newline, start, self.current_pos(), TokenFlags::NONE)
    }

    fn lex_whitespace(&mut self, start: usize) -> Token {
        while let Some(c) = self.peek_char() {
            if c.is_whitespace() && c != '\n' && c != '\r' {
                self.advance();
            } else {
                break;
            }
        }
        self.create_token(TokenKind::Whitespace, start, self.current_pos(), TokenFlags::NONE)
    }

    fn lex_code_block(&mut self, start: usize) -> Token {
        for _ in 0..3 { self.advance(); }
        self.create_token(TokenKind::CodeBlock, start, self.current_pos(), TokenFlags::NONE)
    }

    fn lex_inline_code(&mut self, start: usize) -> Token {
        self.advance();
        self.create_token(TokenKind::InlineCode, start, self.current_pos(), TokenFlags::NONE)
    }

    fn lex_line_comment(&mut self, start: usize) -> Token {
        while let Some(c) = self.peek_char() {
            if c == '\n' || c == '\r' { break; }
            self.advance();
        }
        self.create_token(TokenKind::Comment, start, self.current_pos(), TokenFlags::NONE)
    }

    fn lex_block_comment(&mut self, start: usize) -> Token {
        self.advance(); // consume '/'
        self.advance(); // consume '*'
        let mut prev_star = false;
        while let Some(c) = self.peek_char() {
            self.advance();
            if prev_star && c == '/' {
                break;
            }
            prev_star = c == '*';
        }
        self.create_token(TokenKind::Comment, start, self.current_pos(), TokenFlags::NONE)
    }

    fn lex_string(&mut self, start: usize, quote_type: char) -> Token {
        self.advance();
        let mut escaped = false;

        while let Some(c) = self.peek_char() {
            self.advance();
            if escaped {
                escaped = false;
                continue;
            }
            if c == '\\' {
                escaped = true;
            } else if c == quote_type {
                break;
            }
        }
        self.create_token(TokenKind::String, start, self.current_pos(), TokenFlags::NONE)
    }

    fn lex_number(&mut self, start: usize) -> Token {
        let mut has_decimal = false;
        while let Some(c) = self.peek_char() {
            if c.is_ascii_digit() {
                self.advance();
            } else if c == '.' && !has_decimal {
                let mut temp_chars = self.chars.clone();
                if let Some((_, next_c)) = temp_chars.next() {
                    if next_c.is_ascii_digit() {
                        has_decimal = true;
                        self.advance();
                        continue;
                    }
                }
                break;
            } else {
                break;
            }
        }
        self.create_token(TokenKind::Number, start, self.current_pos(), TokenFlags::NONE)
    }

    fn lex_url(&mut self, start: usize) -> Token {
        while let Some(c) = self.peek_char() {
            if c.is_whitespace() || "()[]{}<>\"'".contains(c) {
                break;
            }
            self.advance();
        }
        self.create_token(TokenKind::Url, start, self.current_pos(), TokenFlags::NONE)
    }

    fn lex_word(&mut self, start: usize) -> Token {
        let mut flags = TokenFlags::NONE;
        if let Some(c) = self.input[start..].chars().next() {
            if c.is_uppercase() {
                flags.insert(TokenFlags::CAPITALIZED);
            }
        }
        let mut all_caps = true;
        let mut has_numbers = false;

        while let Some(c) = self.peek_char() {
            if c.is_alphanumeric() || c == '_' || c == '-' || c == '\'' {
                if c.is_numeric() {
                    has_numbers = true;
                }
                if c.is_lowercase() {
                    all_caps = false;
                }
                self.advance();
            } else {
                break;
            }
        }

        if all_caps && (self.current_pos() - start > 1) {
            flags.insert(TokenFlags::ALL_CAPS);
        }
        if has_numbers {
            flags.insert(TokenFlags::HAS_NUMBERS);
        }

        self.create_token(TokenKind::Word, start, self.current_pos(), flags)
    }

    fn create_token(&self, kind: TokenKind, start: usize, end: usize, flags: TokenFlags) -> Token {
        Token { kind, start, end, flags }
    }
}

File: compressor/prompt_compressor_py/prompt_compressor/lexer.py
"""
Tokenizer / lexer.

Direct port of the Rust `lexer.rs`. A Token never owns text -- it's a
(kind, start, end, flags) tuple over byte offsets into the original
string, exactly like the Rust version. That keeps memory flat no matter
how large the prompt is (no per-token string allocation).
"""

from __future__ import annotations

from enum import IntEnum, IntFlag, auto
from typing import List, NamedTuple, Optional


class TokenKind(IntEnum):
    WORD = auto()
    NUMBER = auto()
    STRING = auto()
    URL = auto()
    CODE_BLOCK = auto()
    INLINE_CODE = auto()
    PUNCTUATION = auto()
    OPERATOR = auto()
    BRACKET = auto()
    WHITESPACE = auto()
    NEWLINE = auto()
    COMMENT = auto()
    UNKNOWN = auto()


class TokenFlags(IntFlag):
    NONE = 0
    CAPITALIZED = 1 << 0
    ALL_CAPS = 1 << 1
    HAS_NUMBERS = 1 << 2
    IS_STOP_WORD = 1 << 3
    IS_BOILERPLATE = 1 << 4
    # Set on tokens synthesized by a compression pass (e.g. a single
    # collapsed space standing in for a run of whitespace) rather than
    # sliced directly from the source text.
    SYNTHETIC = 1 << 5


class Token(NamedTuple):
    kind: TokenKind
    start: int
    end: int
    flags: TokenFlags

    def text(self, source: str) -> str:
        return source[self.start:self.end]


_BRACKETS = set("()[]{}<>")
_PUNCT = set(".,;:!?")
_OPERATORS = set("+-*/=~|&^%\\")


class Lexer:
    """Single-pass scanner over `input`. Mirrors lexer.rs closely."""

    __slots__ = ("input", "n", "pos", "prev_char")

    def __init__(self, text: str):
        self.input = text
        self.n = len(text)
        self.pos = 0
        self.prev_char: Optional[str] = None

    def lex_all(self) -> List[Token]:
        tokens: List[Token] = []
        tok = self.next_token()
        while tok is not None:
            tokens.append(tok)
            tok = self.next_token()
        return tokens

    # -- helpers -----------------------------------------------------

    def _peek(self) -> Optional[str]:
        return self.input[self.pos] if self.pos < self.n else None

    def _starts_with(self, prefix: str) -> bool:
        return self.input.startswith(prefix, self.pos)

    def _advance(self):
        if self.pos < self.n:
            self.prev_char = self.input[self.pos]
            self.pos += 1
        else:
            self.prev_char = None

    def _tok(self, kind: TokenKind, start: int, flags: TokenFlags = TokenFlags.NONE) -> Token:
        return Token(kind, start, self.pos, flags)

    # -- main dispatch -------------------------------------------------

    def next_token(self) -> Optional[Token]:
        if self.pos >= self.n:
            return None
        start = self.pos
        c = self.input[self.pos]

        if c == '\n' or c == '\r':
            return self._lex_newline(start)
        if c.isspace():
            return self._lex_whitespace(start)

        if c == '`':
            if self._starts_with('```'):
                return self._lex_code_block(start)
            return self._lex_inline_code(start)

        if c in _BRACKETS:
            self._advance()
            return self._tok(TokenKind.BRACKET, start)

        if c in _PUNCT:
            self._advance()
            return self._tok(TokenKind.PUNCTUATION, start)

        if c == '/':
            if self._starts_with('//'):
                return self._lex_line_comment(start)
            if self._starts_with('/*'):
                return self._lex_block_comment(start)

        if c in _OPERATORS:
            self._advance()
            return self._tok(TokenKind.OPERATOR, start)

        if c == '"':
            return self._lex_string(start, '"')

        if c == "'":
            prev_is_alnum = self.prev_char is not None and self.prev_char.isalnum()
            if prev_is_alnum:
                self._advance()
                return self._tok(TokenKind.PUNCTUATION, start)
            return self._lex_string(start, "'")

        if c.isdigit():
            return self._lex_number(start)

        if c.isalpha() or c == '_':
            if self._starts_with('http://') or self._starts_with('https://'):
                return self._lex_url(start)
            return self._lex_word(start)

        self._advance()
        return self._tok(TokenKind.UNKNOWN, start)

    # -- sub-lexers ------------------------------------------------------

    def _lex_newline(self, start: int) -> Token:
        while True:
            c = self._peek()
            if c == '\n' or c == '\r':
                self._advance()
            else:
                break
        return self._tok(TokenKind.NEWLINE, start)

    def _lex_whitespace(self, start: int) -> Token:
        while True:
            c = self._peek()
            if c is not None and c.isspace() and c != '\n' and c != '\r':
                self._advance()
            else:
                break
        return self._tok(TokenKind.WHITESPACE, start)

    def _lex_code_block(self, start: int) -> Token:
        for _ in range(3):
            self._advance()
        return self._tok(TokenKind.CODE_BLOCK, start)

    def _lex_inline_code(self, start: int) -> Token:
        self._advance()
        return self._tok(TokenKind.INLINE_CODE, start)

    def _lex_line_comment(self, start: int) -> Token:
        while True:
            c = self._peek()
            if c is None or c == '\n' or c == '\r':
                break
            self._advance()
        return self._tok(TokenKind.COMMENT, start)

    def _lex_block_comment(self, start: int) -> Token:
        self._advance()  # '/'
        self._advance()  # '*'
        prev_star = False
        while True:
            c = self._peek()
            if c is None:
                break
            self._advance()
            if prev_star and c == '/':
                break
            prev_star = (c == '*')
        return self._tok(TokenKind.COMMENT, start)

    def _lex_string(self, start: int, quote: str) -> Token:
        self._advance()
        escaped = False
        while True:
            c = self._peek()
            if c is None:
                break
            self._advance()
            if escaped:
                escaped = False
                continue
            if c == '\\':
                escaped = True
            elif c == quote:
                break
        return self._tok(TokenKind.STRING, start)

    def _lex_number(self, start: int) -> Token:
        has_decimal = False
        while True:
            c = self._peek()
            if c is None:
                break
            if c.isdigit():
                self._advance()
            elif c == '.' and not has_decimal:
                nxt = self.input[self.pos + 1] if self.pos + 1 < self.n else None
                if nxt is not None and nxt.isdigit():
                    has_decimal = True
                    self._advance()
                    continue
                break
            else:
                break
        return self._tok(TokenKind.NUMBER, start)

    def _lex_url(self, start: int) -> Token:
        stop_chars = set("()[]{}<>\"'")
        while True:
            c = self._peek()
            if c is None or c.isspace() or c in stop_chars:
                break
            self._advance()
        return self._tok(TokenKind.URL, start)

    def _lex_word(self, start: int) -> Token:
        flags = TokenFlags.NONE
        first = self.input[start]
        if first.isupper():
            flags |= TokenFlags.CAPITALIZED

        all_caps = True
        has_numbers = False

        while True:
            c = self._peek()
            if c is not None and (c.isalnum() or c in "_-'"):
                if c.isnumeric():
                    has_numbers = True
                if c.islower():
                    all_caps = False
                self._advance()
            else:
                break

        if all_caps and (self.pos - start > 1):
            flags |= TokenFlags.ALL_CAPS
        if has_numbers:
            flags |= TokenFlags.HAS_NUMBERS

        return self._tok(TokenKind.WORD, start, flags)
